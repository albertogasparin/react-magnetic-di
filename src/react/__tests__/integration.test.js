/* eslint-env jest */

import React, { Component, Fragment } from 'react';
import { mount } from 'enzyme';
import { di, DiProvider, mock } from '../../index';

const Wrapper = ({ children }) => children;
const Text = () => 'original';

class Label extends Component {
  render() {
    const [_Wrapper, _Text] = di([Wrapper, Text], Label);
    return (
      <_Wrapper>
        <_Text />
      </_Wrapper>
    );
  }
}

class Input extends Component {
  render() {
    const [_Text] = di([Text], Input);
    return <_Text />;
  }
}

const TextMock = mock(Text, () => 'mock');
const WrapperMock = mock(Wrapper, ({ children }) => children);

describe('Integration', () => {
  it('should return real dependencies if provider less', () => {
    const wrapper = mount(
      <Fragment>
        <Label />
        <Input />
      </Fragment>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should override all dependencies of same type', () => {
    const wrapper = mount(
      <DiProvider use={[TextMock]}>
        <Label />
        <Input />
      </DiProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should allow override composition', () => {
    const wrapper = mount(
      <DiProvider use={[WrapperMock]}>
        <DiProvider use={[TextMock]}>
          <Label />
          <Input />
        </DiProvider>
      </DiProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should only override dependencies of specified target', () => {
    const wrapper = mount(
      <DiProvider target={[Input]} use={[WrapperMock]}>
        <DiProvider target={Label} use={[TextMock]}>
          <Label />
          <Input />
        </DiProvider>
      </DiProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });
});
