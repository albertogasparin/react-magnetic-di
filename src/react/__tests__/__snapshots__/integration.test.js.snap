// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration should allow override composition 1`] = `
"<DiProvider use={{...}}>
  <DiProvider use={{...}}>
    <Label>
      <di(Wrapper)>
        <di(Text)>
          replacement
        </di(Text)>
      </di(Wrapper)>
    </Label>
    <Input>
      <di(Text)>
        replacement
      </di(Text)>
    </Input>
  </DiProvider>
</DiProvider>"
`;

exports[`Integration should get closest dependency if multiple providers using same type 1`] = `
"<DiProvider use={{...}}>
  <Label>
    <Wrapper>
      <di(Text)>
        replacement
      </di(Text)>
    </Wrapper>
  </Label>
  <withDi(Input)>
    <DiProvider use={{...}} target={{...}}>
      <Input>
        <di(Text)>
          closest injectable
        </di(Text)>
      </Input>
    </DiProvider>
  </withDi(Input)>
</DiProvider>"
`;

exports[`Integration should only override dependencies of specified target 1`] = `
"<DiProvider target={{...}} use={{...}}>
  <DiProvider target={[Function: Label]} use={{...}}>
    <Label>
      <Wrapper>
        <di(Text)>
          replacement
        </di(Text)>
      </Wrapper>
    </Label>
    <Input>
      <Text>
        original
      </Text>
    </Input>
  </DiProvider>
</DiProvider>"
`;

exports[`Integration should override all dependencies of same type 1`] = `
"<DiProvider use={{...}}>
  <Label>
    <Wrapper>
      <di(Text)>
        replacement
      </di(Text)>
    </Wrapper>
  </Label>
  <Input>
    <di(Text)>
      replacement
    </di(Text)>
  </Input>
</DiProvider>"
`;

exports[`Integration should return real dependencies if provider less 1`] = `
"<Label>
  <Wrapper>
    <Text>
      original
    </Text>
  </Wrapper>
</Label>


<Input>
  <Text>
    original
  </Text>
</Input>"
`;
